File: python/algorithm/hw5/LIS.py
'''
@Description:
@Author: HuYi
@Date: 2020-04-04 23:15:39
@LastEditors: HuYi
@LastEditTime: 2020-04-06 22:44:46
'''
import math


def bin_search(last, cnt, w):
    left = 0
    right = cnt
    while (left <= right):
        mid = int((left + right) / 2)
        if (last[mid] > w):
            right = mid - 1
        elif (last[mid] < w):
            left = mid + 1
        else:
            return mid
    return left


def cal(a):
    N = len(a)
    cnt = 0
    last = []
    last.append(a[0])
    for i in range(1, N):
        if (a[i] > last[cnt]):
            cnt += 1
            last.append(a[i])
        else:
            location = bin_search(last, cnt, a[i])
            last[location] = a[i]
    length = cnt+1
    last.append(math.inf)
    lis = [0]*length
    for i in range(N-1, -1, -1):
        if (last[cnt] <= a[i] < last[cnt+1]):
            lis[cnt] = a[i]
            cnt -= 1
        if (cnt == -1):
            break
    print('The length of LIS is:', length)
    print('LIS:', lis)


print('Please enter the sequence, separated by only one space, ended by enter')
a = input()
n = (len(a)+1)/2
array = [int(n) for n in a.split()]
cal(array)
os.system("pause")

File: C/algorithm/hw5/LIS.cpp
/*
 * @Description: 
 * @Author: HuYi
 * @Date: 2020-04-06 11:02:27
 * @LastEditors: HuYi
 * @LastEditTime: 2020-04-06 11:07:34
 */

#include <iostream>
#include <vector>
#include <time.h>
using namespace std;

#define N 10
void printSequence(int *b, int *nums, int last);
int main()
{
    srand((unsigned)time(NULL));
    int nums[N + 1] = {0};
    cout << "原序列长度为" << N + 1 << "，如下：" << endl;
    for (int i = 0; i <= N; i++)
    {
        nums[i] = rand() % 1000 + 1;
        cout << nums[i] << " ";
    }

    //数组b存储在递增子序列中当前元素的前一个元素序号
    int b[N + 1] = {0};             //数组b存储在递增子序列中当前元素的前一个元素序号,根据数组b我们能回溯整个子序列
    int last[N + 1] = {0}, L = 0;   //last中存储各不同长度的递增子序列(同一长度的子序列，只考虑尾元素最小的序列)中最后一个元素的序号,根据last数组我们能找到最长子序列的最后一个元素序号。
    for (int i = 1; i < N + 1; i++) //所以我们用last数组找到MAX子序列最后一个元素下标后，利用数组b回溯整个MAX子序列。

    {
        int low = 0, high = L - 1;
        if (nums[i] > nums[last[L == 0 ? 0 : high]])
        { //如果当前元素比last中所有元素都大则最长递增子序列的长度加一，其尾元素为当前元素nums[i]
            last[L++] = i;
            b[i] = last[high];
        }
        else
        { //否则使用二分法在last中查找到大于等于当前元素nums[i]的最小元素
            while (low <= high)
            {
                int middle = (high + low) / 2;
                if (nums[i] > nums[last[middle]])
                    low = middle + 1;
                else
                    high = middle - 1;
            }
            //更新last中的元素值和b中的下标值。
            last[low] = i;
            b[i] = last[low == 0 ? 0 : low - 1];
        }
    }
    int len = L;
    cout << endl;
    cout << "公共子序列如下:" << endl;
    printSequence(b, nums, last[len - 1]);
    cout << endl;
    return 0;
}
void printSequence(int *b, int *nums, int last)
{ //回溯法
    if (b[last] != last)
        printSequence(b, nums, b[last]);
    cout << nums[last] << " ";
}